gini impurity of dataset

In [35]:
total = 14
yes_count = 9
no_count = 5

p_yes = yes_count / total
p_no= no_count / total


gini_impurity_dataset = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for Dataset : {gini_impurity_dataset:.4f}")

Gini Impurity for Dataset : 0.4592


## Gini Impurity for Dataset : 0.4592

- Outlook

In [36]:
# Sunny
total = 5
yes_count = 2
no_count = 3

p_yes = yes_count / total
p_no= no_count / total


sunny_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for Sunny : {sunny_imp:.4f}")

Gini Impurity for Sunny : 0.4800


In [37]:
# overcast
total = 4
yes_count = 4
no_count = 0

p_yes = yes_count / total
p_no= no_count / total


overcast_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for overcast : {overcast_imp:.4f}")

Gini Impurity for overcast : 0.0000


In [38]:
# Rain
total = 5
yes_count = 3
no_count = 2

p_yes = yes_count / total
p_no= no_count / total


Rain_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for Rain : {Rain_imp:.4f}")

Gini Impurity for Rain : 0.4800


In [39]:
GI_outlook = (0.48*5+0.00*4+0.48*5)/14
GI_outlook

0.34285714285714286

## Gini impurity for Outlook = 0.3428

- Temperature


In [40]:
# hot
total = 4
yes_count = 2
no_count = 2

p_yes = yes_count / total
p_no= no_count / total


hot_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for hot : {hot_imp:.4f}")

Gini Impurity for hot : 0.5000


In [41]:
# mild
total = 6
yes_count = 4
no_count = 2

p_yes = yes_count / total
p_no= no_count / total


mild_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for hot : {mild_imp:.4f}")

Gini Impurity for hot : 0.4444


In [42]:
# cool
total = 4
yes_count = 3
no_count = 1

p_yes = yes_count / total
p_no= no_count / total


mild_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for cool : {mild_imp:.4f}")

Gini Impurity for cool : 0.3750


In [43]:
GI_Temp = (0.50*4+0.44*6+0.37*4)/14
GI_Temp

0.4371428571428572

## Gini impurity for Temperature = 0.4371

- Humidity

In [44]:
# High
total = 7
yes_count = 3
no_count = 4

p_yes = yes_count / total
p_no= no_count / total


sunny_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for High : {sunny_imp:.4f}")

Gini Impurity for High : 0.4898


In [45]:
# Normal
total = 7
yes_count = 6
no_count = 1

p_yes = yes_count / total
p_no= no_count / total


sunny_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for Normal : {sunny_imp:.4f}")

Gini Impurity for Normal : 0.2449


In [46]:
GI_Humidity = (0.49*7+0.24*7)/14
GI_Humidity

0.36499999999999994

## GI impurity for Humidity = 0.365

- wind

In [47]:
# weak
total = 8
yes_count = 6
no_count = 2

p_yes = yes_count / total
p_no= no_count / total


sunny_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for weak : {sunny_imp:.4f}")

Gini Impurity for weak : 0.3750


In [48]:
# strong
total = 6
yes_count = 3
no_count = 3

p_yes = yes_count / total
p_no= no_count / total


sunny_imp = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for strong : {sunny_imp:.4f}")

Gini Impurity for strong : 0.5000


In [49]:
GI_Wind = (0.37*8+0.50*6)/14
GI_Wind

0.4257142857142857

## The gini impurity for wind is 0.426

so the first nide will be outlook, and the second node will be humidity, and the third node will be wind

Advantages of Decision Trees:

Interpretability: Decision trees are easy to understand and interpret. They represent decisions and their consequences in a clear and intuitive manner, making them suitable for explaining the reasoning behind a prediction.

No Assumptions about Data: Decision trees do not make assumptions about the data distribution, making them applicable to a wide range of data types, including categorical and numerical features.

Handling Non-linearity: Decision trees can model complex, non-linear relationships in the data, which may not be captured by linear models.

Feature Selection: Decision trees naturally perform feature selection by identifying the most important features at the top of the tree. This helps in identifying relevant features in the dataset.

Robustness to Outliers: Decision trees are relatively robust to outliers, as they split the data based on thresholds and can isolate extreme values in separate branches.

Ease of Handling Missing Data: Decision trees can handle missing data by using surrogate splits, which allow them to make predictions even when some values are missing.

Scalability: Decision trees can handle large datasets efficiently and are relatively fast to train, especially when compared to complex algorithms like neural networks.

Disadvantages of Decision Trees:

Overfitting: Decision trees are prone to overfitting, especially when the tree is deep and overly complex. This can lead to poor generalization on unseen data.

Instability: Small variations in the data can lead to different tree structures, making decision trees less stable compared to some other algorithms.

Bias Towards Dominant Classes: Decision trees tend to favor features with more levels or values, which can lead to a bias toward dominant classes in classification problems.

Difficulty in Modeling Some Relationships: Complex relationships may require very deep trees, which can be hard to interpret and prone to overfitting.

Lack of Continuity: Decision trees create step-like boundaries, which can be problematic for problems where continuous predictions or smooth decision boundaries are required.

Sensitive to Data Imbalance: If the classes in a classification problem are imbalanced, decision trees may not perform well, as they tend to favor the majority class.

Greedy Nature: Decision trees use a greedy, top-down approach for splitting data, which may not always lead to the globally optimal tree structure.

Difficulty with XOR-Like Problems: Decision trees struggle with problems like XOR, where the decision boundaries are more complex and involve combinations of features.

High Memory Usage: Deep decision trees can be memory-intensive, especially when dealing with a large number of features and nodes.

### Consider a binary classification problem with two classes: "spam" and "not spam". You have a dataset of 500 emails, of which 200 are spam and 300 are not spam. Calculate the Gini index for this dataset.

In [50]:
total = 700
yes_count = 200
no_count = 300

p_yes = yes_count / total
p_no= no_count / total


gini_impurity_dataset = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for Dataset : {gini_impurity_dataset:.4f}")

Gini Impurity for Dataset : 0.7347


### You are training a decision tree model on a dataset of 1000 customer records, where each record has two features: age (in years) and income (in thousands of dollars), and the target variable is "customer churn" (yes or no). You decide to split the data on the age feature using a threshold of 30 years. After splitting the data, you have 600 records in the left subtree and 400 records in the right subtree. Of the 600 records in the left subtree, 400 are non-churners and 200 are churners. Of the 400 records in the right subtree, 150 are non-churners and 250 are churners. Calculate the Gini index for each of the two subtrees.

In [51]:
left_subtree_tot = 600
left_subtree_yes = 200
left_subtree_no = 400
p_yes = left_subtree_yes / left_subtree_tot
p_no= left_subtree_no / left_subtree_tot


gini_impurity_dataset = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for left_sub_tree : {gini_impurity_dataset:.4f}")

Gini Impurity for left_sub_tree : 0.4444


In [52]:
right_subtree_tot = 400
right_subtree_yes = 250
right_subtree_no = 150
p_yes = right_subtree_yes / right_subtree_tot
p_no= right_subtree_no / right_subtree_tot


gini_impurity_dataset = 1 - (p_yes ** 2 + p_no ** 2)

print(f"Gini Impurity for right_sub_tree : {gini_impurity_dataset:.4f}")

Gini Impurity for right_sub_tree : 0.4688


In [60]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

### Given a dataset of children with kyphosis(a spinal condition) who underwent an operation. The dataset has the following columns:

- Kyphosis: whether or not the kyphosis was present after the operation.
- Age: age of children in months
- Number: number of vertebrae involved in the operation
- Start: the topmost vertebrae number involved in the operation.
Use 70% of the data for training and the rest for testing. Use Decision Tree Classifier model and make predictions for the test data. Also, evaluate the model.

In [61]:
df=pd.read_csv('datasets\Assignment -10 dataset - Sheet1.csv')
df.head()

Kyphosis  Age  Number  Start
0   absent   71       3      5
1   absent  158       3     14
2  present  128       4      5
3   absent    2       5      1
4   absent    1       4     15

In [62]:
desc=pd.DataFrame(index=list(df))
desc['count']=df.shape[0]
desc['unique']=df.nunique()
desc['null']=df.isnull().sum()
desc['dtype']=df.dtypes
desc = pd.concat([desc,df.describe().T],axis=1)
desc

count  unique  null   dtype  count       mean        std  min   25%  \
Kyphosis     81       2     0  object    NaN        NaN        NaN  NaN   NaN   
Age          81      64     0   int64   81.0  83.654321  58.104251  1.0  26.0   
Number       81       8     0   int64   81.0   4.049383   1.619423  2.0   3.0   
Start        81      16     0   int64   81.0  11.493827   4.883962  1.0   9.0   

           50%    75%    max  
Kyphosis   NaN    NaN    NaN  
Age       87.0  130.0  206.0  
Number     4.0    5.0   10.0  
Start     13.0   16.0   18.0

In [63]:
for i in df.columns:
    print(df[i].unique())

['absent' 'present']
[ 71 158 128   2   1  61  37 113  59  82 148  18 168  78 175  80  27  22
 105  96 131  15   9   8 100   4 151  31 125 130 112 140  93  52  20  91
  73  35 143  97 139 136 121 177  68  72 120  51 102 114  81 118  17 195
 159 127  87 206  11 178 157  26  42  36]
[ 3  4  5  2  6  7  9 10]
[ 5 14  1 15 16 17 12  2 18 13  9  3  6 11 10  8]


In [64]:
df['Kyphosis'] = df['Kyphosis'].map({'absent':0,'present':1})

In [65]:
x=df.drop('Kyphosis',axis=1)
y = df['Kyphosis']

In [66]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=42)

In [67]:
model = DecisionTreeClassifier()
model.fit(x_train,y_train)

DecisionTreeClassifier()

In [68]:
y_pred = model.predict(x_test)
y_pred

array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 0], dtype=int64)

In [69]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [74]:
cm = confusion_matrix(y_test,y_pred)
TP = cm[0][0]
TN = cm[1][1]
FP = cm[0][1]
FN = cm[1][0]
print(TP,TN,FP,FN)

15 1 4 5


In [70]:
ac = accuracy_score(y_test,y_pred)
ac

0.64

In [73]:
rprt = classification_report(y_test,y_pred)
print(rprt)

              precision    recall  f1-score   support

           0       0.75      0.79      0.77        19
           1       0.20      0.17      0.18         6

    accuracy                           0.64        25
   macro avg       0.47      0.48      0.48        25
weighted avg       0.62      0.64      0.63        25



since false negatives are 5 and false positives are 4 the model is not good